satellite/satellitedb/nodeselection.go

```diff
@@ -0,0 +1,327 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package satellitedb
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/zeebo/errs"
+
+	"storj.io/common/pb"
+	"storj.io/common/storj"
+	"storj.io/private/version"
+	"storj.io/storj/satellite/overlay"
+)
+
+func (cache *overlaycache) SelectStorageNodes(ctx context.Context, totalNeededNodes, newNodeCount int, criteria *overlay.NodeCriteria) (nodes []*overlay.SelectedNode, err error) {
+	defer mon.Task()(&ctx)(&err)
+	if totalNeededNodes == 0 {
+		return nil, nil
+	}
+
+	if newNodeCount > totalNeededNodes {
+		return nil, Error.New("requested new node count can't exceed requested total node count")
+	}
+
+	needNewNodes := newNodeCount
+	needReputableNodes := totalNeededNodes - needNewNodes
+
+	receivedNewNodes := 0
+	receivedNodeNetworks := make(map[string]struct{})
+
+	var excludedIDs []storj.NodeID
+	excludedIDs = append(excludedIDs, criteria.ExcludedIDs...)
+
+	var excludedNetworks []string
+	excludedNetworks = append(excludedNetworks, criteria.ExcludedNetworks...)
+
+	for i := 0; i < 3; i++ {
+		reputableNodes, newNodes, err := cache.selectStorageNodesOnce(ctx, needReputableNodes, needNewNodes, criteria, excludedIDs, excludedNetworks)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, node := range newNodes {
+			// checking for last net collision among reputable and new nodes since we can't check within the query
+			if _, ok := receivedNodeNetworks[node.LastNet]; ok {
+				continue
+			}
+			excludedIDs = append(excludedIDs, node.ID)
+			excludedNetworks = append(excludedNetworks, node.LastNet)
+			nodes = append(nodes, node)
+			needNewNodes--
+			receivedNewNodes++
+
+			if criteria.DistinctIP {
+				receivedNodeNetworks[node.LastNet] = struct{}{}
+			}
+		}
+		for _, node := range reputableNodes {
+			if _, ok := receivedNodeNetworks[node.LastNet]; ok {
+				continue
+			}
+			excludedIDs = append(excludedIDs, node.ID)
+			excludedNetworks = append(excludedNetworks, node.LastNet)
+			nodes = append(nodes, node)
+			needReputableNodes--
+
+			if criteria.DistinctIP {
+				receivedNodeNetworks[node.LastNet] = struct{}{}
+			}
+		}
+
+		if needNewNodes > 0 && receivedNewNodes == 0 {
+			needReputableNodes += needNewNodes
+			needNewNodes = 0
+		}
+
+		if needReputableNodes <= 0 && needNewNodes <= 0 {
+			break
+		}
+	}
+	return nodes, nil
+}
+
+func (cache *overlaycache) selectStorageNodesOnce(ctx context.Context, reputableNodeCount, newNodeCount int, criteria *overlay.NodeCriteria, excludedIDs []storj.NodeID, excludedNetworks []string) (reputableNodes, newNodes []*overlay.SelectedNode, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	newNodeSelection := `SELECT last_net, id, address, last_ip_port, true FROM nodes`
+	reputableNodeSelection := `SELECT last_net, id, address, last_ip_port, false FROM nodes`
```

Comment:
whats the false about here?

Reply:
when the query is run, it returns both new and reputable nodes together, and I wasn't sure of an easy way to distinguish the two. the bool here represents `isNewNode`- so when we scan the rows later, we can scan this value into a bool to determine what kind of node it is. it's weird. I can add a comment for clarification.